### ISSUE TITLE
Add a wiki?

### ISSUE THREAD

**Issue by appleguru** [Dilbert66/esphome-vistaECP#11]
Any way you can add a wiki to this project so we can build up a knowledge base?  (Eg, compatible hardware, examples builds, BOM…)

**Comment by Dilbert66** [comment]
I really don't have the resources at this time to maintain it and have not seen any real requests for more info then what is already available. It's a fairly basic circuit.  If there is some confusion , I can always amend the docs.

**Comment by appleguru** [comment]
Just set it up and open up permissions so anyone can edit? Happy to contribute with a build guide/some photos/getting started/etc.

I’m still trying to sort out “ideal” part numbers for the voltage regulator (which if I understand correctly should allow it to be powered from the vista system?), for the optocoupler, and for the nodeMCU/ESP module itself. I think I can order everything I need for the build from mouser (if I’m willing to overpay a bit for the esp module).

**Comment by Dilbert66** [comment]
https://www.amazon.ca/Converter-Module-Keenso-Step-Down-Conversion/dp/B07PQLNXVC/ref=sr_1_26?dchild=1&keywords=step+down+mini&qid=1615644230&sr=8-26

or any lm2596 based regulator
https://www.amazon.ca/Converter-Ajustable-Step-down-Converters-1-25V-35V/dp/B074J4DLWQ/ref=asc_df_B074J4DLWQ/?tag=googleshopc0c-20&linkCode=df0&hvadid=335365451913&hvpos=&hvnetw=g&hvrand=1334836585404895701&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=9000757&hvtargid=pla-899437512675&psc=1

Both of those work extremely well.  None of the parts are critical. Any inexpensive optocoupler will work. Same goes with the esp module.  Any inexpensive one will work. I do suggest though that you power it with 3.3 volts (at the 3.3 volt input of course) instead of with 5 volts to avoid using the weak built in 5volt regulators on some esp modules. This is why i recommend the regulators above (or similar) as they are fully adjustable. 

As to the wiki, I'll think about it.

**Comment by appleguru** [comment]
What about for an optocoupler? Will a cheap PC817 work?

**Comment by Dilbert66** [comment]
Sure, I think it would work perfectly

**Comment by appleguru** [comment]
Thanks. Did an Amazon order (which I generally avoid for electronics due to all the counterfeit parts, but should be ok for this circuit…)

Thanks for putting this together, excited to finally tie my (to date completely useless) existing security system into home assistant!

**Comment by appleguru** [comment]
Ok, my build is done and seems to be (mostly?) working. I think the remaining issues are likely due to configuration which I could definitely use some help with.

![DEC75D14-0C74-4ACC-97BC-399E31722FE1](https://user-images.githubusercontent.com/595829/111373828-c1c15a00-8672-11eb-8f0a-335655c49d1a.jpeg)
![3B79820C-DF07-4517-A1D0-13CCEC429AB2](https://user-images.githubusercontent.com/595829/111373830-c259f080-8672-11eb-9f52-d71c1613a889.jpeg)
<img width="243" alt="038AF06B-F62E-49FB-9D49-1125E64CE2AC" src="https://user-images.githubusercontent.com/595829/111373912-de5d9200-8672-11eb-9490-9f0cdd3d5ce4.png">

I disabled the expander, relay, and lrrsupervisor in the config since I *think* I have real devices on the system for all of those at present.

Arming/disarming from home assistant also don’t seem to be working (but this is likely user error? I’m trying to use Developer Tools -> Services for this now but it seems to have no effect. I put my alarm code into a secrets.yaml file as “access_code”

I think I also might be missing some devices? (But am not 100% sure what I have on the vista system). Any help would be greatly appreciated!

**Comment by Dilbert66** [comment]
If you can't send keys that means your tx circuit is not working. Looking at your circuit soldering, the optocoupler looks to be miswired.  It looks like your are bridging pins 3 and 4.

**Comment by appleguru** [comment]
I don’t think the pins are bridged, but will double check with a meter when I get home. What’s the easiest way to test TX? Arm/disarm?

**Comment by Dilbert66** [comment]
Looking at the soldering, they do look soldered together.  only pins 2 and 3 go to ground.
![image](https://user-images.githubusercontent.com/7193213/111379986-28964180-867a-11eb-989b-ea9571f73061.png)

**Comment by Dilbert66** [comment]
Arm /disarm/send a key. You should see a logged response on the esphome log window.

**Comment by appleguru** [comment]
The circled pins are actually 2/3 on the options:
![9E5A2E34-A058-4DE7-8A50-91DC040E650A](https://user-images.githubusercontent.com/595829/111384494-f4be1a80-867f-11eb-838c-2043afdf9b0a.jpeg)

I think this might have my current alarm code in it, so will probably delete or edit after, but this is what I see for debug when I try to arm:
<img width="1427" alt="96BCE6D5-124E-4C93-8789-B258DE2C45BE" src="https://user-images.githubusercontent.com/595829/111384641-20d99b80-8680-11eb-889e-854ff9c000c4.png">

**Comment by Dilbert66** [comment]
If i were you I would double check the pinout on the 817 

https://www.twinschip.com/PC817-Optocoupler

**Comment by appleguru** [comment]
> If i were you I would double check the pinout on the 817 . The wiring is all wrong from what I see.
> 
> https://www.twinschip.com/PC817-Optocoupler

I think i have it right? (I misspoke before, I circled 3/4, not 2/3)

**Comment by Dilbert66** [comment]
Yes, that's correct but on the picture pin2 is going direct to  the esp module? I don't get it? That should be ground

**Comment by appleguru** [comment]
> Yes, that's correct but on the picture pin2 is going direct to the esp module? I don't get it? That should be ground

It’s going to a ground pin on the ESP module. They all should be connected; was just the closest ground to the part.

**Comment by Dilbert66** [comment]
ah ok..  I see now.  The solder on the underside threw me off also.

**Comment by Dilbert66** [comment]
Either way, if the log is not showing  a key was sent then, you might still have a wiring issue somewhere.

**Comment by appleguru** [comment]
> Either way, if the log is not showing a key was sent then, you might still have a wiring issue somewhere.

Can you take a look at the log screenshot above? I’m not sure what I’m looking at/if it shows the key being sent (I think it does?)

**Comment by Dilbert66** [comment]
It looks like your sending fine now that I have a closer look at your log image.  The F6 is receiving from device id 0x10 (address 16) which I assume is your esp.

Try using the alarm_keypress service from developer tools and send any keys for example. You can send any keys you want there.

service: esphome.vistaalarmtest_alarm_keypress
data: {"keys":"*1234*1234"}

**Comment by Dilbert66** [comment]
This is my test system so I'm using 1234 as an access code

**Comment by Dilbert66** [comment]
Those F6 ext cmds (the EXT is a pseudo field to display sent keys from peripherals in response to the previous F6 CMD) don't make sense.  they show a lenght of 1 for the associated. Normally for an arm cmd you will see a lenght of 6 ie. the 4 digit code, the 1 digit arm code and a 1 digit checksum.   That would explain why it's not arming.

**Comment by adamurban** [comment]
Interesting.. Any idea what might cause that? Here's the output using the same 12342 example that you posted (I also of course tried with my actual alarm code too).

<img width="1419" alt="Screen Shot 2021-03-16 at 7 29 47 PM" src="https://user-images.githubusercontent.com/3276234/111393067-4e2d4600-868e-11eb-814c-817202adba7e.png">

**Comment by Dilbert66** [comment]
Interesting. This looks to be a software bug somewhere.  I'm running the same version as the master I think.  I'll check to see.

**Comment by Dilbert66** [comment]
I recompiled my test version from the master copy and everything works fine on my end.  Well, it's a mystery to me why your version is behaving like that. I assume you are using the latest master code version. I really can't see how the issue you have would occur.  
What panel version do have?

**Comment by Dilbert66** [comment]
Can you post a longer copy and paste of the logs so I can view normal interactions.   Also can you enter any random sequence of characters in the physical keypad and post the associated esp logs as well.

**Comment by Dilbert66** [comment]
It's possible that's it's conflicting with another device. Can you change your esp address to another address such as 17 instead of the default 16.

**Comment by adamurban** [comment]
Trying a different address now; if that doesn't solve it I'll grab some logs.

**Comment by adamurban** [comment]
Success! Moved to address 17 and arming/disarming works ok now.

**Comment by appleguru** [comment]
Success! Moved to address 17 and arming/disarming works ok now.

**Comment by appleguru** [comment]
Panel is version 10.23, which is why I assume address 17 worked for a keypad…
Looking at the hardware, I have:
1) Vista-20p control panel
2) 6160 keypad
3) 5881ENL wireless receiver
4) GSMV4G alarm radio (for remote access??)
5) SRN2000 motion sensor
6) a crap load of contact and bar sensors…
7) WI-MTA1/5817 wireless temp sensor
8) WI-WB200 water sensor
9) ADEMCO-4204 ECP relay module (connected to a system sensor COSMOD2W 2wire fire alarm interface (which connects to COSMO2W smoke/CO detectors)
10) SDM-100 siren driver connected to a AS-SHX51 outdoor siren
11) 4219 wired zone expander
12) MPI35F indoor siren

And probably some other things that I missed. I don’t currently have the programming code for the system, but am hoping to get that tomorrow.

Any tips for identifying everything in HA/making sure I’m not missing any devices?

**Comment by adamurban** [comment]
From DIP switches, I see that my real 4219 wired zone expander is on address 7 and my real 4204 relay module is on address 12.

Presumably because I have a GSMV4G, I want lrrSupervisor: "false" right?
And if I want to use relay and/or zone emulation, I would want to change the default relayAddr1: "12" to 13 or something else, but expanderAddr1: "8" is OK. (I have these all set to 0 for now as I don't yet have a need for the emulated inputs).

Is there a good way to figure out what addresses everything else in the alarm system are set to?

**Comment by appleguru** [comment]
From DIP switches, I see that my real 4219 wired zone expander is on address 7 and my real 4204 relay module is on address 12.

Presumably because I have a GSMV4G, I want lrrSupervisor: "false" right?
And if I want to use relay and/or zone emulation, I would want to change the default relayAddr1: "12" to 13 or something else, but expanderAddr1: "8" is OK. (I have these all set to 0 for now as I don't yet have a need for the emulated inputs).

Is there a good way to figure out what addresses everything else in the alarm system are set to?

**Comment by Dilbert66** [comment]
That's quite a loaded system you have. Yes, using the virtual expander with address 8 is ok and changing the relayaddr1 to 13 is correct also. And yes, since you already have a GSMv4g, you definitively want to set lrrSupervisor to False.  That interface is used to report faults to a remote using cell or pager.  With my system I just use those messages for display as it provides more info.

As to finding out what zones you have, there are two ways I guess.  Open up every zone and record the message displayed on the keypad or the esp.  The other method requires having the installer code and going to program *56, press 1 (to confirm) then use the * key to step through all the programming to see what zones are setup and what type. 

Did you try installer code 4112 or 6321 . Those are the defaults .

You enter 4112800 or 6321800 and see if it gets you the installer prompt.   You should download a vista20p installer manual . Just google it. Lots of copies around.

**Comment by appleguru** [comment]
Yeah, the alarm company changed the default installer code; I tried both of those. I called them today and they said they should be able to remotely reset it back to the default code for me tomorrow; we’ll see if it works?

Used your example template to add a keypad, looks nice! I don’t really need/want to have to enter the code each time to disarm, so just hard-coded the “code” field and manually added the “disarm” state, which seems to work. Only problem now is the virtual keypad is useless… any way to make it pass the button presses through to the alarm_keypress service so I can use it like the real keypad on the wall?

**Comment by Dilbert66** [comment]
What's needed is a different virtual keypad component. You are right that the existing keypad is very limited as it is now. Not sure what's available but that's a good idea though to fully emulate a keypad. If there isnt one around, I might try my hand at coding one up.  

Once you find out your zones, you'll need to add new sections in the yaml to represent those and edit the existing to match your setup.

**Comment by appleguru** [comment]
I don’t think it quite works like this, but being able to conditionally call a service might work… something like:

`
        disarm: >
            {% if {{code}} %}
              - service: esphome.vistaalarm_alarm_keypress
                data_template:
                  keys: '{{code}}'
            {% else %}
              - service: esphome.vistaalarm_alarm_disarm
              data_template:
                code: '1234'
            {% endif %}
        `

**Comment by Dilbert66** [comment]
Maybe, I have not really looked too deeply into it since as it is the default panel is limited in it's ability to work as a standalone keypad since it is missing the "#" and "*" keys which are needed in a lot of commands, such as programming, etc.   I'm going to look into adapting an existing custom card to add those

**Comment by appleguru** [comment]
Ok, progress.. found some system documentation with zone numbers listed that seems mostly correct, and also got the programming code.


 I updated the yaml template entires and case statement entries like this:

```
   - platform: template
    id: z15
    name: "$systemName Basement Water Sensor"
    

case 15: id(z15).publish_state(open); break;
```

I now have a bunch on unavailable entities that I can’t seem to remove for the zones I no longer have, but the things that are there seem to be working. Will dive into the panel config on the keypad later and double check everything once I figure out how :D

One question: I see a few entries in the log like this:
```
 Got relay address 7 channel 12 = 0
 Got relay address 12 channel 56 = 0
 Got relay address 12 channel 53 = 0
 ```
Does this mean there are input devices I don’t yet have mapped? Or can I ignore these? I guess I still need to wrap my head around how addressing works in the vista system (address vs channel vs zone…)

**Comment by Dilbert66** [comment]
You might need to delete the esphome vista alarm entity and re-add it. That's the only I found to remove ghost entries after changes.  Those relay messages are status changes from your panel.  You can ignore those unless you want to see the relay activities.  You can also comment out that log message from vistaalarm.h if you don't want to see it. It's more of a debug message than anything.

**Comment by Dilbert66** [comment]
By the way, I'm testing a custom keypad component that will work just like a regular alarm keypad with all keys. I also found an issue with the way my code handles key writes that I need to fix up so I will be pushing an update in a day or so.

**Comment by Dilbert66** [comment]
I've pushed a new commit to the "DEV" branch.  This fixes the key send issue and adds a new custom panel card which emulates a regular keypad.   You'll need to use the new yaml or at least add the changes below:
```
    VistaECP->onLine1DisplayChange([&](std::string msg) {
        id(l1).publish_state(msg); 
    });  
    VistaECP->onLine2DisplayChange([&](std::string msg) {
        id(l2).publish_state(msg); 
    });   
``
```
  - platform: template
    id: l1
    name: "$systemName Line1"
 

  - platform: template
    id: l2
    name: "$systemName Line2"
```

**Comment by Dilbert66** [comment]
I will add a few more buttons later such as the arm night/day/etc buttons and disarm.

**Comment by adamurban** [comment]
Nice work! Giving it a shot now

**Comment by adamurban** [comment]
I think the c files are missing a function (unless I copied the wrong files??)

```
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
src/main.cpp: In lambda function:
src/main.cpp:616:17: error: 'class vistaECPHome' has no member named 'onLine1DisplayChange'
       VistaECP->onLine1DisplayChange([&](std::string msg) {
                 ^
src/main.cpp:619:17: error: 'class vistaECPHome' has no member named 'onLine2DisplayChange'
       VistaECP->onLine2DisplayChange([&](std::string msg) {
                 ^
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/WiFiClient.cpp.o
*** [/data/vistaalarm/.pioenvs/vistaalarm/src/main.cpp.o] Error 1
```

**Comment by appleguru** [comment]
I think the c files are missing a function (unless I copied the wrong files??)

```
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
src/main.cpp: In lambda function:
src/main.cpp:616:17: error: 'class vistaECPHome' has no member named 'onLine1DisplayChange'
       VistaECP->onLine1DisplayChange([&](std::string msg) {
                 ^
src/main.cpp:619:17: error: 'class vistaECPHome' has no member named 'onLine2DisplayChange'
       VistaECP->onLine2DisplayChange([&](std::string msg) {
                 ^
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling /data/vistaalarm/.pioenvs/vistaalarm/lib4d9/ESP8266WiFi/WiFiClient.cpp.o
*** [/data/vistaalarm/.pioenvs/vistaalarm/src/main.cpp.o] Error 1
```

**Comment by Dilbert66** [comment]
My copy missed some files. I've pushed them now. Try it again.

**Comment by Dilbert66** [comment]
I've updated the card to now allow configuring the 4 function keys to send any sequence of keys.  You will need to update the resource file like so: /local/alarm-keypad-card-js?ver=1 (where 1 will increment everytime you need to refresh changes to it), You will then need to do an F5 on the browser to refresh the page and load the new version.

**Comment by appleguru** [comment]
Slick!

I didn’t see your example card yaml anywhere, so hard to type that from your screenshot… here’s mine:

```
type: 'custom:alarm-keypad-card'
title: Vista
unique_id: vista
kpd_line1: sensor.vistaalarm_line1
kpd_line2: sensor.vistaalarm_line2
scale: 1
view_pad: true
kpd_service: sensor.vistaalarm_keypress
button_A: STAY
button_B: AWAY
button_C: DISARM
button_D: BYPASS
cmd_A: '12343'
cmd_B: '12342'
cmd_C: '12341'
cmd_D: '12346#'
```

<img width="1340" alt="BB3164FA-A2D2-46E3-8BC9-21751A515ABA" src="https://user-images.githubusercontent.com/595829/111682982-5bb40e80-87fb-11eb-9089-263b2ef44ba8.png">

**Comment by appleguru** [comment]
This virtual keypad is absolute 🔥… I can now do programming right from HA and don’t have to stand in the hallway taking photos of the LCD like a dork!

**Comment by Dilbert66** [comment]
I didnt post a yaml since I configured it using the gui. I can post a yaml too

**Comment by appleguru** [comment]
> I didnt post a yaml since I configured it using the gui. I can post a yaml too

Maybe I missed the UI option? How/where would I find that? (I didn’t see it in “Add Card”, clicked custom at the bottom and just typed from your example screenshot. It helpfully prompted me for missing fields)

**Comment by Dilbert66** [comment]
When you edit a dashboard, you will then see an Edit on each card where you can make changes there.

**Comment by appleguru** [comment]
> When you edit a dashboard, you will then see an Edit on each card where you can make changes there.

right, but where does the card come from in the first place?

**Comment by Dilbert66** [comment]
from the js code in the resources. When you add a card, it just load the js code and displays it in whatever dashboard you are using.

**Comment by Dilbert66** [comment]
I've pushed an update that fixes an issue with the * being sent while in programming mode.  This would cause odd behaviour if your trying to program the system while using the card.

**Comment by appleguru** [comment]
Thanks, seems to be working well. I noticed you checked in some mp3 files and made a few changes for "beeps". Do those work? I copied them over but they don't seem to do anything.

Unrelated, I opened a small PR for the js to improve the button size/spacing to make the keypad behave better on mobile (and also match the native themes): https://github.com/Dilbert66/esphome-vistaECP/pull/12

**Comment by Dilbert66** [comment]
The beep is still a work in progress.  It needs to have the beep duration better controlled.  It does  beep depending on the beep code sent from the esp (you need to also have the beep section setup in the esp yaml).
Add this line to the card config:
```
beep: sensor.vistaalarmtest_beeps
```

Awesome about the PR, I'll check it out. I also want to add feedback on a keypress to show visually when the button is pressed.  This polymer entity stuff is new to me .

**Comment by Dilbert66** [comment]
Tried the changes. Excellent. Exactly what it needed.

**Comment by appleguru** [comment]
I had to enable sound "auto play" in my browser too (it worked ok in the HA app out of the box at least).

Those beeps are pretty bad; maybe I'll just make some recordings of the actual beeps..

**Comment by appleguru** [comment]
Some more things to eventually go in the wiki if we make one:

Sensor entities from this integration can return the following states:
O = Open / Fault
C = Closed
A = Alarm
T = Trouble
B = Bypass

because they have more than two states, they aren't binary sensors, and the built in "Color icons based on state?" toggle (`state_color`) will have no impact.

There are lots of ways to address this, but I opted to use custom-ui. You can download a version compatible with the latest Home Assistant here: https://github.com/Mariusthvdb/custom-ui

Once you install the custom-ui.js file, you can add an icon-color glob rule to /config/customize_glob.yaml like this:

```
sensor.vistaalarm_*:
  templates:
    icon_color: >
      if (state == 'C') return 'steelblue';
      return 'var(--paper-item-icon-active-color)';
```

Make sure you also enable the glob .yaml by adding this to your main configuration.yaml if you don't already have it:

``` 
customize_glob: !include customize_glob.yaml
```

**Comment by Dilbert66** [comment]
Good info. Tks.

**Comment by AlexandreUSA** [comment]
Just found this thread. I will start the project soon to connect my Vista 21ip to HA. Can't wait to get the electronics on order.

**Comment by appleguru** [comment]
I see we have a new "Discussions" area on this project. Will start moving some of this information there, since GitHub wiki's aren't really community editable.

https://github.com/Dilbert66/esphome-vistaECP/discussions

Also, gave the latest code a whirl... I like it a lot! Having more buttons I can program is helpful (Eg, buttons to enter/exit programming mode since I'm a Vista n00b and always have to google it otherwise), and I like the status "lights". Only real (minor) nit now is that the bottom row of buttons doesn't auto align with the rest of the virtual panel (likely because they're in their own DIV?).

<img width="342" alt="Screen Shot 2021-05-21 at 12 39 38 AM" src="https://user-images.githubusercontent.com/595829/119083043-3e531a80-b9cd-11eb-8bdf-6be10fb543f8.png">

**Comment by Dilbert66** [comment]
The idea was to be able to hide the bottom row since some users might not need it which is why I used a separate div but it causes the alignment issue you noticed.  For the vista the <> buttons are not needed for navigation but they are used on the DSC which is why I have them there. I wanted a keypad that was flexible enough to use on different systems.    I didnt spend too much time on the style to try and fix it as I am more concerned with the functions for now.  I'm modifying my dsc application to fully emulate a DSC LCD keypad.

**Comment by Dilbert66** [comment]
> Some more things to eventually go in the wiki if we make one:
> 
> Sensor entities from this integration can return the following states:
> O = Open / Fault
> C = Closed
> A = Alarm
> T = Trouble
> B = Bypass
> 
> because they have more than two states, they aren't binary sensors, and the built in "Color icons based on state?" toggle (`state_color`) will have no impact.
> 
> There are lots of ways to address this, but I opted to use custom-ui. You can download a version compatible with the latest Home Assistant here: https://github.com/Mariusthvdb/custom-ui
> 
> Once you install the custom-ui.js file, you can add an icon-color glob rule to /config/customize_glob.yaml like this:
> 
> ```
> sensor.vistaalarm_*:
>   templates:
>     icon_color: >
>       if (state == 'C') return 'steelblue';
>       return 'var(--paper-item-icon-active-color)';
> ```
> 
> Make sure you also enable the glob .yaml by adding this to your main configuration.yaml if you don't already have it:
> 
> ```
> customize_glob: !include customize_glob.yaml
> ```
> 
> That's it, once you reboot your entities should be colored when not in the closed state, like this:
> 
> <img alt="Screen Shot 2021-03-28 at 6 22 35 PM" width="365" src="https://user-images.githubusercontent.com/595829/112770048-9d5e6980-8ff2-11eb-8bd0-0d851152f10a.png">


I'm going to add this to the readme when I get a chance.

**Comment by bmwhitetx** [comment]
Sorry to open an old discussion but speaking of the onscreen display I thought I would show a quick mod I made to the file **alarm-keypad-card.js**. I didn't like dull blue background color and wanted to match the display screen with the backlight on (used [this site](https://www.w3schools.com/colors/colors_picker.asp) to pick color).  Here is the change I made and what it looks like:

        .keypad_display {
          background: #94ff4d;

![image](https://github.com/Dilbert66/esphome-vistaECP/assets/60900216/259b4382-02c1-4baa-b553-7ab0330ad5b1)

Thanks for a great program. I have mine working using the first hardware schematic.

**Comment by bmwhitetx** [comment]
Yes, definitely a preference, just wanted to show an option - it was fun trying to figure out the code and how to install the panel. Speaking of which, the instructions on adding this interface panel in the readme.md somehow is tagged as code so it was difficult to read (no line wrap). I ended up just copying the whole section and pasting into an editor with word wrap and was able to follow the instructions. I think the part about the hardware interfaces got duplicated in there too. Its the part that starts with:
'##  Setting up the alarm panel keyboard card on HA

And while you're in there, HA, as they are prone to do, moved things around - so this part:
Using UI: Configuration → Lovelace Dashboards → Resources Tab →  Click Plus button → ...
Should now be:
Using UI: Settings → Dashboards → Click three dots in upper right → Choose Resources →  Click Add Resource button → ...
The file option may have changed as well, I didn't have a lovelace section in my configuration.yaml file.

Thanks again for making this project!
